src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -381,52 +391,48 @@ DropShards(Oid relationId, char *schemaName, char *relationName,
 		CoordinatedTransactionUse2PC();
 	}
 
-	ShardInterval *shardInterval = NULL;
-	foreach_ptr(shardInterval, deletableShardIntervalList)
+	List *dropTaskList = DropTaskList(relationId, schemaName, relationName,
+									  deletableShardIntervalList);
+
+	Task *task = NULL;
+	foreach_ptr(task, dropTaskList)
 	{
-		uint64 shardId = shardInterval->shardId;
+		uint64 shardId = task->anchorShardId;
 		char *shardRelationName = pstrdup(relationName);
 
-		Assert(shardInterval->relationId == relationId);
-
 		/* build shard relation name */
 		AppendShardIdToName(&shardRelationName, shardId);
 
-		List *shardPlacementList = ShardPlacementList(shardId);
-
 		ShardPlacement *shardPlacement = NULL;
-		foreach_ptr(shardPlacement, shardPlacementList)
+		foreach_ptr(shardPlacement, task->taskPlacementList)
 		{
 			uint64 shardPlacementId = shardPlacement->placementId;
+			int32 shardPlacementGroupId = shardPlacement->groupId;
 
-			if (shardPlacement->groupId == COORDINATOR_GROUP_ID &&
-				IsCoordinator() &&
-				DropSchemaOrDBInProgress())
+			bool isLocalShardPlacement = (shardPlacementGroupId == localGroupId);
+
+			if (isLocalShardPlacement && IsCoordinator() && DropSchemaOrDBInProgress())
 			{
 				/*
 				 * The active DROP SCHEMA/DATABASE ... CASCADE will drop the
 				 * shard, if we try to drop it over another connection, we will
-				 * get into a distributed deadlock.
+				 * get into a distributed deadlock. Hence, just delete the shard
+				 * placement metadata and skip it for now.
 				 */
+				DeleteShardPlacementRow(shardPlacementId);
+				continue;
 			}
-			else
-			{
-				char storageType = shardInterval->storageType;
-
-				const char *dropShardPlacementCommand =
-					CreateDropShardPlacementCommand(schemaName, shardRelationName,
-													storageType);
 
-				/*
-				 * Try to open a new connection (or use an existing one) to
-				 * connect to target node to drop shard placement over that
-				 * remote connection
-				 */
-				ExecuteDropShardPlacementCommandRemotely(shardPlacement,
-														 shardRelationName,
-														 dropShardPlacementCommand);
+			if (isLocalShardPlacement)
```

Comment:
very minor: If you can move this below `ExecuteDropShardPlacementCommandRemotely`, it'd be slightly easier to read